File: Ruby/stack-sort.rb
# This algorithm is to sort a stack in place, using no additional data structures

# Define stack class

class Stack
  def initialize
    @stack = []
  end

  def push(data)
    @stack.push(data)
  end

  def pop
    @stack.pop
  end

  def peek
    @stack[@stack.length - 1]
  end

  def is_empty?
    @stack.length == 0
  end

  def to_s
    stack_string = @stack.map { |item| "#{item}, " }.join("")
    print stack_string[0..-3] + "\n"
  end
end

# Write method to sort stack

def sort_stack(stack)
  if !stack.is_empty?
    holder = stack.pop
    sort_stack(stack)
    insert_sorted(stack, holder)
  end
  return stack
end

def insert_sorted(stack, item)
  if stack.peek == nil || stack.peek < item
    stack.push(item)
  else
    holder = stack.pop
    insert_sorted(stack, item)
    stack.push(holder)
  end
end

# Create stack

stack = Stack.new

stack.push(5)
stack.push(10)
stack.push(2)
stack.push(10)
stack.push(78)
stack.push(7)

# Call method on stack

puts sort_stack(stack)

File: Javascript/stack-sort.js
// Define a stack constructor function

function Stack() {
  this.stack = [];

  this.push = function(data) {
    this.stack.push(data);
  }

  this.pop = function() {
    return this.stack.pop();
  }

  this.peek = function() {
    return this.stack[this.stack.length - 1];
  }

  this.isEmpty = function() {
    return this.stack.length === 0;
  }
}

// Define functions to sort stack in place

function sortStack(stack) {
  if (!stack.isEmpty()) {
    var element = stack.pop();
    sortStack(stack);
    insertSorted(stack, element);
  }
  return stack.stack;
}

function insertSorted(stack, data) {
  if (stack.isEmpty() || stack.peek() < data) {
    stack.push(data);
  } else {
    var holder = stack.pop();
    insertSorted(stack, data);
    stack.push(holder);
  }
}

// Create a stack

var stack = new Stack;

stack.push(10);
stack.push(7);
stack.push(12);
stack.push(8);
stack.push(9);

// Call sort stack function

console.log(sortStack(stack));